File: crates/kimi-protocol/src/usage.rs
//! Usage wire types — moved from `packages/kimi-agent/src/usage/mod.rs`
//! (Rust-first migration, stage A3). Pure serde types; the engine's
//! `UsageRecorder` stays in kimi-core and re-exports these.

use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use crate::wire_types::TokenUsage;

/// Per-model usage summary.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UsageStatus {
    /// Usage broken down by model alias.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub by_model: Option<HashMap<String, TokenUsage>>,
    /// Total usage across all models.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub total: Option<TokenUsage>,
    /// Usage for the current turn (if any).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub current_turn: Option<TokenUsage>,
}

File: packages/agent-core-v2/src/agent/usage/usage.ts
import type { AgentContext } from '#/agent/agentContext/agentContext';
import type { AgentLLMRequestSource } from '#/actor/llmRequester/llmRequester';
import type { TokenUsage } from '#/kosong/contract/usage';

export interface UsageStatus {
  readonly byModel?: Record<string, TokenUsage>;
  readonly total?: TokenUsage;
  readonly currentTurn?: TokenUsage;
}

export interface UsageRecordedContext {
  readonly agent: AgentContext;
  readonly model: string;
  readonly usage: Readonly<TokenUsage>;
  readonly source?: AgentLLMRequestSource;
  readonly firstRecord: boolean;
}
